### PR TITLE
Better support for OpenBSD

### PR DESCRIPTION
**Pull request by qbit**
## Description

This PR adds support for parsing `man.conf`, interfaces and packages on OpenBSD. It also brings in a set of completions for OpenBSD tools.

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: share/completions/pfctl.fish [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+complete -c pfctl -s s --description 'Show a filter parameter by modifyer' -xa 'queue rules Anchors states Sources info labesl timeouts memory Tables ospf Interfaces all'
+complete -c pfctl -s F --description 'Flush the filter params specified by mod' -xa 'rules states Sources info Tables ospf all'
+complete -c pfctl -s T --description 'Table command' -xa 'kill flush add delete expire replace show test zero'
```
Comment (faho) [inline share/completions/pfctl.fish]: More descriptions.

File: share/completions/rcctl.fish [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+complete -c rcctl -xa 'check ls reload restart stop start disable enable' -n 'not __fish_seen_subcommand_from list check ls reload restart stop start enable disable'
+complete -c rcctl -n '__fish_seen_subcommand_from check reload restart stop start enable disable' -xa '( ls /etc/rc.d )'
```
Comment (faho) [inline share/completions/rcctl.fish]: Instead of `ls`, use `printf`. I.e. `(set -l files /etc/rc.d/*; printf "%s\n" $files)`.

File: share/completions/signify.fish [hunk 3]
```diff
@@ -0,0 +1,4 @@
+complete -c signify -n '__fish_use_subcommand' -s C --description 'Verify a signed checksum list'
```
Comment (faho) [inline share/completions/signify.fish]: `__fish_use_subcommand` ignores things that start with a dash, so this would have no effect.

Comment (faho) [reply share/completions/signify.fish]: Use `__fish_seen_subcommand_from`.

File: share/completions/vmctl.fish [hunk 4]
```diff
@@ -0,0 +1,4 @@
+
+complete -c vmctl -xa 'console create load log reload reset start status stop pause unpause send receive' -n 'not __fish_seen_subcommand_from list console create load log reload reset start status stop pause unpause send receive'
+complete -c vmctl -n '__fish_seen_subcommand_from console reload reset start status stop pause unpause send receive' -xa '( vmctl status | grep -v NAME | awk \'{print $NF}\' )'
```
Comment (faho) [inline share/completions/vmctl.fish]: Preferably, this'd use `string`, since it's a builtin. The `grep` is simple - `string match -v`. I never really learned awk - what does the output look like and what are you trying to get out?

File: share/functions/__fish_print_interfaces.fish [hunk 5]
```diff
@@ -3,6 +3,11 @@ function __fish_print_interfaces --description "Print a list of known network in
         set -l interfaces /sys/class/net/*
         string replace /sys/class/net/ '' $interfaces
     else # OSX/BSD
-        command ifconfig -l | string split ' '
+        set -l os (uname)
+        if string match -q "OpenBSD" -- $os
```
Comment (faho) [inline share/functions/__fish_print_interfaces.fish]: Does the openbsd output here differ from the macOS/FreeBSD stuff?

What does it look like? Can we make one path that works for both?

File: share/functions/__fish_print_interfaces.fish [hunk 6]
```diff
@@ -3,6 +3,11 @@ function __fish_print_interfaces --description "Print a list of known network in
...
+            command ifconfig | grep -e "^[a-z]" | awk -F: '{print $1}' | string split ' '
```
Comment (faho) [inline share/functions/__fish_print_interfaces.fish]: Again, the grep can be trivially replaced with `string match -r "^[a-z]"`. The `awk` should be `string replace -r ':.*' ''`.

File: share/functions/__fish_print_packages.fish [hunk 7]
```diff
@@ -35,6 +35,17 @@ function __fish_print_packages
         return
     end
 
+    # pkg_info on OpenBSD provides versioning info which we want for
+    # installed packages
+    if begin
```
Comment (faho) [inline share/functions/__fish_print_packages.fish]: This `begin` and `end` is unnecessary since about 2.3.0. Now `and` and similar bind to `if`.

File: share/functions/__fish_print_packages.fish [hunk 8]
```diff
@@ -35,6 +35,17 @@ function __fish_print_packages
...
+            type -q -f pkg_add
+            and test (uname) = "OpenBSD"
```
Comment (faho) [inline share/functions/__fish_print_packages.fish]: I like not using `uname` if we can somehow get away with it. Since we currently have no path for `pkg_add`, can we just remove this condition?

File: share/functions/__fish_print_packages.fish [hunk 9]
```diff
@@ -35,6 +35,17 @@ function __fish_print_packages
...
+        end
+	# pkg_info can get blocked if there is another instance of pkg_* running.
+        ls -1 /var/db/pkg
```
Comment (faho) [inline share/functions/__fish_print_packages.fish]: I'd again replace the `ls` with `set -l files /var/db/pkg/*` and `printf '%s\n' $files`.

File: share/functions/__fish_print_packages.fish [hunk 10]
```diff
@@ -35,6 +35,17 @@ function __fish_print_packages
         return
     end
 
+    # pkg_info on OpenBSD provides versioning info which we want for
+    # installed packages
+    if begin
+            type -q -f pkg_add
+            and test (uname) = "OpenBSD"
+        end
+	# pkg_info can get blocked if there is another instance of pkg_* running.
```
Comment (faho) [inline share/functions/__fish_print_packages.fish]: Is this why you're not calling `pkg_info` here? That should probably be stated outright.

File: share/tools/create_manpage_completions.py [hunk 11]
```diff
@@ -866,15 +866,25 @@ def get_paths_from_manpath():
...
-            try:
-                proc = subprocess.Popen(prog, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-            except OSError: # Command does not exist, keep trying
-                continue
-            break # Command exists, use it.
-        manpath, err_data = proc.communicate()
-        parent_paths = manpath.decode().strip().split(':')
+        if os.uname()[0] == 'OpenBSD':
+            parent_paths = ['/usr/share/man', '/usr/X11R6/man', '/usr/local/man']
+            if os.path.isfile('/etc/man.conf'):
```
Comment (faho) [inline share/tools/create_manpage_completions.py]: Does anyone else use another incompatible /etc/man.conf?

If nobody does, we can just try this after we've tried to call the tools.

File: share/functions/__fish_print_packages.fish (hunk 9 above)
Comment (qbit) [reply share/functions/__fish_print_packages.fish]: Is there a "fish" way of removing the path from that output? `%s\n` ends up having `/var/db/pkg/yquake2-7.10` - which will break pkg_* tools.

File: share/functions/__fish_print_packages.fish (hunk 8 above)
Comment (qbit) [reply share/functions/__fish_print_packages.fish]: IMO there needs to be some way to distinguish between pkg_* variants - for example pkg_add exists in pkgsrc and FreeBSD (this might be replaced with pkg by now.. not entirely sure.)

File: share/tools/create_manpage_completions.py (hunk 11 above)
Comment (qbit) [reply share/tools/create_manpage_completions.py]: I am not entirely sure if anyone else uses it.

File: share/completions/pfctl.fish [hunk 12]
```diff
@@ -0,0 +1,4 @@
+
+complete -c pfctl -s s --description 'Show a filter parameter by modifyer' -xa 'queue rules Anchors states Sources info labesl timeouts memory Tables ospf Interfaces all'
```
[the remark this answers is not in the document]
Comment (qbit) [reply share/completions/pfctl.fish]: Do you mean more description in the `--description` section? or a comment for each option describing a bit more?

File: share/functions/__fish_print_interfaces.fish (hunk 5 above)
Comment (qbit) [reply share/functions/__fish_print_interfaces.fish]: It does, ya.. Might be possible to make one that works for both.. I am not well versed in 'fish' thought.. Here is the output of a typical system:

```
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 32768
        index 4 priority 0 llprio 3
        groups: lo
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x4
        inet 127.0.0.1 netmask 0xff000000
em0: flags=8802<BROADCAST,SIMPLEX,MULTICAST> mtu 1500
        lladdr ff:ff:ff:ff:ff:ff
        index 1 priority 0 llprio 3
        media: Ethernet autoselect (none)
        status: no carrier
iwm0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
        lladdr ff:ff:ff:ff:ff:ff
        index 2 priority 4 llprio 3
        groups: wlan egress
        media: IEEE802.11 autoselect (HT-MCS13 mode 11n)
        status: active
        ieee80211: nwid WiFi chan 1 bssid ff:ff:ff:ff:ff:ff 100% wpakey <not displayed> wpaprotos wpa2 wpaakms psk wpaciphers ccmp wpagroupcipher ccmp
        inet 255.255.255.255 netmask 0xffffff00 broadcast 255.255.255.255
```

File: share/completions/vmctl.fish (hunk 4 above)
Comment (qbit) [reply share/completions/vmctl.fish]:
```
vmctl status
   ID   PID VCPUS  MAXMEM  CURMEM     TTY        OWNER NAME
    1     -     1    512M       -       -         qbit ssb
    2     -     1    1.0G       -       -         qbit chrome
    3     -     1    512M       -       -         qbit alpine
    4     -     1    512G       -       -         qbit void
```

File: share/functions/__fish_print_interfaces.fish (hunk 5 above)
Comment (qbit) [reply share/functions/__fish_print_interfaces.fish]: I should also mention there is no `-l` on OpenBSD.

File: share/functions/__fish_print_interfaces.fish (hunk 6 above)
Comment (qbit) [reply share/functions/__fish_print_interfaces.fish]: looks like the `string match -r "^[a-z]"` part only returns the first character of each match.. is there a way to make it print the entire line when the RE is matched?

File: share/completions/pfctl.fish (hunk 12 above)
[the remark this answers is not in the document]
Comment (faho) [reply share/completions/pfctl.fish]: A comment for each option. You can either use a function like

```fish
function __fish_pfctl_filter_modifiers
    printf '%s\t%s\n' queue "something to describe queue"
    printf '%s\t%s\n' rules "something to describe rules"
    # ....
end
```

or something like

```fish
complete -c pfctl -s s --description 'Show a filter parameter by modifyer' -xa \
'queue\t"something to describe queue"
rules\t"something for rules"'
```

File: share/functions/__fish_print_interfaces.fish (hunk 5 above)
Comment (faho) [reply share/functions/__fish_print_interfaces.fish]: Look at the example output [in the FreeBSD quickstart guide](https://www.freebsd.org/doc/en/articles/linux-users/network.html):

```
% ifconfig
em0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
        options=b<RXCSUM,TXCSUM,VLAN_MTU>
        inet 10.10.10.100 netmask 0xffffff00 broadcast 10.10.10.255
        ether 00:50:56:a7:70:b2
        media: Ethernet autoselect (1000baseTX <full-duplex>)
        status: active
em1: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
        options=b<RXCSUM,TXCSUM,VLAN_MTU>
        inet 192.168.10.222 netmask 0xffffff00 broadcast 192.168.10.255
        ether 00:50:56:a7:03:2b
        media: Ethernet autoselect (1000baseTX <full-duplex>)
        status: active
```

That looks basically like your OpenBSD output. So we could omit the "-l" on both, and just split everywhere.

Note that [the FreeBSD manpage](http://nixdoc.net/man-pages/FreeBSD/ifconfig.8.html) says that "-a" is implied if no other arguments are given. Is that true on OpenBSD as well?

File: share/functions/__fish_print_interfaces.fish (hunk 6 above)
Comment (faho) [reply share/functions/__fish_print_interfaces.fish]: You need to either make the regex match the entire line (`"^[a-z]+$"`) or use the "-e"/"--entire" flag.

Sorry, I forgot about that!

File: share/functions/__fish_print_packages.fish (hunk 8 above)
Comment (faho) [reply share/functions/__fish_print_packages.fish]: Yeah, but we don't support those currently. We do support FreeBSD's `pkg`, so if they still have `pkg_add` and friends, we just need to make sure that `pkg` is checked first.

And if they have `pkg_*`, do they differ in ways that are meaningful here?

File: share/functions/__fish_print_packages.fish (hunk 9 above)
Comment (faho) [reply share/functions/__fish_print_packages.fish]: Ah, yes, I understand now.

Instead of `printf`, use `string replace '/var/db/pkg/' '' -- $files`.

(Note that all this breaks on files with newlines, but we don't support that with completions in general and in practice it simply does not matter)

File: share/tools/create_manpage_completions.py (hunk 11 above)
Comment (faho) [reply share/tools/create_manpage_completions.py]: I can't find anything on centos (which would be Red Hat), but it seems that FreeBSD and macOS use it as well.

See [the FreeBSD manpage](https://www.unix.com/man-page/freebsd/5/man.conf/). Given that that uses a directive called "MANPATH" where [OpenBSD uses "manpath"](http://man.openbsd.org/man.conf.5), we could just try reading either (i.e. "the word manpath, case-insensitively") after we've exhausted the programs (which should still be preferred because they're more likely to get weird edgecases right). If that still doesn't work, we could still fallback to the current default.

As you might have guessed by now, I have a certain philosophy when it comes to cross-platform support, in that I really don't like checking for the specific platform name via `uname` or similar. I very much prefer feature-testing if at all possible. This allows us to e.g. do proper completion for GNU utilities even on systems where they aren't usually installed. (One interesting case where we can't do that is `killall`, because we absolutely cannot invoke it on Solaris - because it actually kills all)

File: share/completions/vmctl.fish (hunk 4 above)
Comment (faho) [reply share/completions/vmctl.fish]: So you want the last field there? Yeah, that's something that isn't quite as nice with `string`.

Try `vmctl status | string match -v 'NAME' | string replace -r '^(\s+\S+\s+){7}' ''`

(Obviously the "NAME" part isn't great because IIUC you could have a VM called "NAME-SOMETHING" - what you'd really want is something like `tail -n +2`, but we can probably live with that)

Comment (qbit) [reply share/completions/vmctl.fish]: This works if i do `string match -e -v`. Also I switched "NAME" to "MAXMEM" - because that's a super lame vm name ;D

**Comment by qbit** [comment]
- [X] Separated into more commits.
- [X] Typo / descriptions.
- [X] Less use of unix utilities in favour of fish built-ins.
- [X] Use `__fish_seen_subcommand_from`.
- [X] Generic ifconfig parsing for BSDs.
- [x] Parse `man.conf` if it exists.

I think I got everything.

File: share/completions/pfctl.fish [hunk 13]
```diff
@@ -0,0 +1,32 @@
+complete -c pfctl -s s --description 'Show a filter parameter by modifier' -xa \
+	'queue\t"Show the currently loaded queue definitions." \
+	rules\t"Show the currently loaded filter rules." \
+	Anchors\t"Show the currently loaded anchors directly attached to the main ruleset." \
```
Comment (faho) [inline share/completions/pfctl.fish]: Descriptions should be as short as possible, so we can ideally fit more columns - `Show anchors attached to main ruleset`?

Also, I prefer no trailing ".".

File: share/completions/pfctl.fish [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+	states\t"Show the contents of the state table."
```
Comment (faho) [inline share/completions/pfctl.fish]: `Show contents of state table` or `Show state table contents`?

File: share/completions/rcctl.fish [hunk 15]
```diff
@@ -0,0 +1,3 @@
+
+complete -c rcctl -xa 'check ls reload restart stop start disable enable' -n 'not __fish_seen_subcommand_from list check ls reload restart stop start enable disable'
+complete -c rcctl -n '__fish_seen_subcommand_from check reload restart stop start enable disable' -xa '(set -l files /etc/rc.d/*; string replace '/etc/rc.d/' '' -- $files)'
```
Comment (faho) [inline share/completions/rcctl.fish]: This probably doesn't work properly because of the quoting - remember, you're already inside single quotes, so if you're going to use them you need to escape them (`\'`). Since that confuses some syntax highlighters, I usually use double-quotes in cases like these. I.e.:

```fish
-xa '(set -l files /etc/rc.d/*; string replace "/etc/rc.d/" "" -- $files)'
```

File: share/completions/vmctl.fish [hunk 16]
```diff
@@ -0,0 +1,4 @@
+
+complete -c vmctl -xa 'console create load log reload reset start status stop pause unpause send receive' -n 'not __fish_seen_subcommand_from list console create load log reload reset start status stop pause unpause send receive'
+complete -c vmctl -n '__fish_seen_subcommand_from console reload reset start status stop pause unpause send receive' -xa '(vmctl status | string match -e -v 'MAXMEM' | string replace -r '^(\s+\S+\s+){7}' '')'
```
Comment (faho) [inline share/completions/vmctl.fish]: Similarly with the quotes. I think those curly braces will actually be stripped (since they are outside of quotes, so fish evaluates them as a brace-expansion - `echo {7}` prints just "7").

File: share/functions/__fish_print_packages.fish [hunk 17]
```diff
@@ -35,6 +35,18 @@ function __fish_print_packages
         return
     end
 
+    # pkg_info on OpenBSD provides versioning info which we want for
+    # installed packages but, calling it directly can cause delays in
+    # returning information if another pkg_* tool have a lock.
+    # Listing /var/db/pkg is a clean alternative.
+    if begin
```
Comment (faho) [inline share/functions/__fish_print_packages.fish]: Again: This `begin; end` pair can be removed.

```fish
if type -q -f pkg_add # which is just `command -sq pkg_add`
   and test (uname) = "OpenBSD"
```

will do the right thing.

I still maintain though that the uname check can be removed as well.

File: share/tools/create_manpage_completions.py [hunk 18]
```diff
@@ -876,10 +875,18 @@ def get_paths_from_manpath():
         manpath, err_data = proc.communicate()
         parent_paths = manpath.decode().strip().split(':')
     if (not parent_paths) or (proc and proc.returncode > 0):
-        # HACK: Use some fallback in case we can't get anything else.
+        # HACK: Use some fallbacks in case we can't get anything else.
         # `mandoc` does not provide `manpath` or `man --path` and $MANPATH might not be set, so just use the default for mandoc (minus /usr/X11R6/man, because that's not relevant).
         # The alternative is reading its config file (/etc/man.conf)
```
Comment (faho) [inline share/tools/create_manpage_completions.py]: This comment needs to be updated - we _are_ reading the mandoc config file.

**Comment by faho** [comment]
> Separated into more commits.

Just the way I wanted it, thanks! Would it be possible to add any remaining changes as one commit per thing as well? Either that or squash them on top of the relevant commits yourself.

>I think I got everything.

Almost.

**Comment by qbit** [comment]
Updated - I was hesitant to remove the uname checks because there was more than just the one for OpenBSD - Now there are none.

**Comment by faho** [comment]
Yes, this seems good to merge. Nice work!

I've just noticed that you've signed all your commits. I'd like to squash them together so we have one commit per "thing", but if I do that I'm going to trample your signatures (or rather I'd create new commit objects that I obviously can't sign with your key).

So would you like to do that to keep the signatures or should I do it?

**Comment by qbit** [comment]
Na, you can go for it :D

**Comment by faho** [comment]
Merged as 277cd305..75e17e00. Thank you very much!
